### PR TITLE
Add price and taxons filters UI for promotions

### PR DESCRIPTION
**Pull request by gabiudrescu**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | yes
| Related tickets | partially #3368
| License         | MIT

with this new addition, you can create campaigns like this:
 - give 10% off for all products that cost more than 100$

if this is well received, the next step is to do another PR with a taxon filter, that will solve the riddle I've posted here: #5772

I couldn't have managed to pull this off without the help of @vcraescu, so much of the kudos goes to him :beers: 

image

image

### CODE REVIEW

**Comment by gabiudrescu** [comment]
I'd also be happy to update the docs from #5946 with more info about how filters should work.

File: src/Sylius/Bundle/PromotionBundle/Form/Type/Filter/ActionFiltersType.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\FormBuilderInterface;
+
+/**
+ * The form type used to filter the promotion subjects when applying an action to them
+ *
+ * @author Viorel Craescu <viorel.craescu@trisoft.ro>
+ * @author Gabi Udrescu <gabriel.udr@gmail.com>
+ */
+
```
Comment (Zales0123) [inline src/Sylius/Bundle/PromotionBundle/Form/Type/Filter/ActionFiltersType.php]: Extra blank line.

File: src/Sylius/Bundle/PromotionBundle/Form/Type/Filter/ActionFiltersType.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
+<?php
+/*
```
Comment (Zales0123) [inline src/Sylius/Bundle/PromotionBundle/Form/Type/Filter/ActionFiltersType.php]: Missing blank line.

File: src/Sylius/Bundle/PromotionBundle/Form/Type/Filter/ActionFiltersType.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\PromotionBundle\Form\Type\Filter;
+
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\FormBuilderInterface;
+
+/**
+ * The form type used to filter the promotion subjects when applying an action to them
```
Comment (Zales0123) [inline src/Sylius/Bundle/PromotionBundle/Form/Type/Filter/ActionFiltersType.php]: Is this comment really necessary?

File: src/Sylius/Bundle/PromotionBundle/Form/Type/Filter/ActionFiltersType.php [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * @author Viorel Craescu <viorel.craescu@trisoft.ro>
+ * @author Gabi Udrescu <gabriel.udr@gmail.com>
+ */
+
+class ActionFiltersType extends AbstractType
+{
+    public function buildForm(FormBuilderInterface $builder, array $options)
```
Comment (Zales0123) [inline src/Sylius/Bundle/PromotionBundle/Form/Type/Filter/ActionFiltersType.php]: Missing docblock.

File: src/Sylius/Bundle/PromotionBundle/Form/Type/Filter/PriceRangeType.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<?php
+/*
```
Comment (Zales0123) [inline src/Sylius/Bundle/PromotionBundle/Form/Type/Filter/PriceRangeType.php]: Missing blank line.

File: src/Sylius/Bundle/PromotionBundle/Form/Type/Filter/PriceRangeType.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace Sylius\Bundle\PromotionBundle\Form\Type\Filter;
+
+use Symfony\Component\Form\AbstractType;
+use Sylius\Bundle\MoneyBundle\Form\Type\MoneyType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Validator\Constraints\Type;
+
+/**
+ * The form type used to filter the promotion subjects using a price range
```
Comment (Zales0123) [inline src/Sylius/Bundle/PromotionBundle/Form/Type/Filter/PriceRangeType.php]: Again, I think it's redundant

File: src/Sylius/Bundle/PromotionBundle/Form/Type/Filter/PriceRangeType.php [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * @author Viorel Craescu <viorel.craescu@trisoft.ro>
+ * @author Gabi Udrescu <gabriel.udr@gmail.com>
+ */
+class PriceRangeType extends AbstractType
+{
+    public function buildForm(FormBuilderInterface $builder, array $options)
```
Comment (Zales0123) [inline src/Sylius/Bundle/PromotionBundle/Form/Type/Filter/PriceRangeType.php]: And again missing docblock.

File: src/Sylius/Bundle/PromotionBundle/Resources/config/services.xml [hunk 8]
```diff
@@ -117,5 +117,15 @@
...
+
+
+        <service id="sylius.form.type.promotion_action.filter" class="Sylius\Bundle\PromotionBundle\Form\Type\Filter\ActionFiltersType">
+            <tag name="form.type" alias="sylius_promotion_action_filter" />
+        </service>
+
+        <service id="sylius.form.type.promotion_action.filter.price_range" class="Sylius\Bundle\PromotionBundle\Form\Type\Filter\PriceRangeType">
+            <tag name="form.type" alias="sylius_promotion_action_filter_price_range" />
+        </service>
+
```
Comment (Zales0123) [inline src/Sylius/Bundle/PromotionBundle/Resources/config/services.xml]: Toooo many blank lines in this change.

**Comment by Zales0123** [comment]
@gabiudrescu I really appreciate your work with filters 👍  I would just suggest to write some Behat scenarios for this feature - we should really cover any additional feature to be 100% about them ;)

**Comment by pjedrzejewski** [comment]
@gabiudrescu Are you willing to try our Behats and write some scenarios for that? I'd be happy to merge it then! :)

**Comment by gabiudrescu** [comment]
@pjedrzejewski sure. I'm planning that by Sunday, the 16th of October, have this in a merge-able form.

I've been a little bit busy lately and I wasn't so sure this PR was good, so last weekend I had a conference call with @vcraescu to better understand his work.

after we talked, I realized that currently, what I have pushed here has a bug - because it also applies for "order discount" - and we want this only for items.

**Comment by gabiudrescu** [comment]
my question about behat: https://github.com/Sylius/Sylius/blob/master/features/promotion/applying_promotion_rules/receiving_discount_based_on_total_of_items_from_specific_taxon.feature#L16 isn't this coupled with a context and ran in TravisCI already?

am I missing something?

**Comment by michalmarcinkowski** [comment]
These scenarios validates that promotion works for order, but after we add UI in the admin panel we need scenarios for creating/editing promotions with filters through UI 😉  (similar to [scenarios for rules](https://github.com/Sylius/Sylius/blob/master/features/promotion/managing_promotions/adding_promotion_with_rule.feature))

**Comment by gabiudrescu** [comment]
fair enough. I think I got now everything I needed so I can start work. thanks for the tip.

File: src/Sylius/Bundle/PromotionBundle/Form/Type/Action/PercentageDiscountConfigurationType.php [hunk 9]
```diff
@@ -36,6 +37,7 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                     new Type(['type' => 'numeric']),
                 ],
             ])
+            ->add('filters', ActionFiltersType::class)
```
Comment (gabiudrescu) [inline src/Sylius/Bundle/PromotionBundle/Form/Type/Action/PercentageDiscountConfigurationType.php]: it is declared as service here: https://github.com/Sylius/Sylius/pull/6054/files#diff-17b78b75b647ab8c629e081f3ee1af1fR122

**Comment by gabiudrescu** [comment]
I have added the taxons filters too, but I haven't had enough time to add behats also.

Will add them in the next session of (pair) programming.

**Comment by gabiudrescu** [comment]
I have added a feature file with a working scenario and some skipped ones.

First of all, are the skipped ones enough or I went too far by testing every possible combination?

Second of all, on my local laptop the scenarios failed because things happened too fast and the success flash messages was checked before the browser was able to render the HTML.  If I made Behat wait a few seconds, it passed.

3rd of all - my Travis run issued some errors for the Sylius\Bundle\CoreBundle\Form\DataTransformer\TaxonsToCodesTransformer class 

```
it is an abstract form
      could not reflect class Sylius\Bundle\CoreBundle\Form\DataTransformer\TaxonsToCodesTransformer as it is marked final.
```

is it OK if I unmark it as not final?

**Comment by pamil** [comment]
@gabiudrescu IMO specs for form types bring us no real value, so I rather vote for removing that spec instead of removing `final` from `TaxonsToCodeTransformer`.

Nevertheless, we shouldn't typehint for (in `ActionFiltersType`) and mock (in `ActionFiltersType`) the `TaxonsToCodeTransformer`, using `TransformerInterface` will allow us to change used transformer in the future and decouple the form type from that form transformer.

**Comment by lchrusciel** [comment]
@gabiudrescu I will need this feature ASAP. Won't you mind if I will finish it for you?

**Comment by gabiudrescu** [comment]
@lchrusciel I won't, but you'll have to allow me to buy you a beer for doing this.

**Comment by lchrusciel** [comment]
@gabiudrescu of course I would not mind :) But firstly, I have to finish all of filters. 

Anyway, can be closed in favour of #6480 /cc @michalmarcinkowski

**Comment by michalmarcinkowski** [comment]
Thanks for your work @gabiudrescu! 👍

**Pull request closed without merging**